roadmap/implementers-guide/src/types/candidate.md

```diff
@@ -159,6 +161,29 @@ struct TransientValidationData {
 	/// vector is sorted ascending by the para id and doesn't contain multiple entries with the same
 	/// sender. This informs the collator.
 	hrmp_mqc_heads: Vec<(ParaId, Hash)>,
+	/// A vector that enumerates the list of blocks in which there was at least one message
+	/// received. The first number is always after the watermark.
+	hrmp_digest: Vec<BlockNumber>,
+	/// The watermark of the HRMP. That is, the block number up to which (inclusive) all HRMP messages
+	/// sent to the parachain are processed.
+	hrmp_watermark: BlockNumber,
```

Comment:
@bkchr One thing to watch out for here, is that because this is in the `TransientValidationData`, it is not a direct parameter to the validation function and instead will be provided by inhernet. That means that a malicious collator can provide something different (e.g. 100 instead of 90) and the relay-chain will still accept the block. So it is possible for this value to move _backwards_ from the parachain's perspective and Cumulus needs to account for that. Similarly with other values in this struct.

Reply:
Actually, after conversation in DM with Sergei, our conclusion was that this watermark should not be provided via inherent and instead Cumulus should be tracking the watermark internally.

Reply:
Not sure I understand this here, why isn't the relay chain checking that the watermark is `>= old_watermark`?

Reply:
The relay-chain actually does check that the watermark is `> old_watermark`.

```
check_hrmp_watermark(P: ParaId, new_hrmp_watermark):
- new_hrmp_watermark should be strictly greater than the value of HrmpWatermarks for P (if any).
- new_hrmp_watermark must not be greater than the context's block number.
- in HrmpChannelDigests for P an entry with the block number equal to new_hrmp_watermark should exist.
```

Reply:
What happens if the parachain can not process all downward messages of a given relay chain block and thus could not increase the value?

Reply:
FWIW, in this particular place, we are discussing the HRMP watermark. For XCMP and by extension HRMP we have agreed on a simplifying assumption: we assume that a parachain can receive and process all the incoming messages within a block. I think that might be fine for HRMP, I don't think that we will reach the usage levels where they could be a problem and we are going to put aggressive limits on HRMP anyway (I see it happening within upcoming #1632).

For XCMP I think we might want to revisit this assumption. Under circumstances when the parachain is bombarded by the stream of messages over all the channels and processing of all these messages just don't fit into the PVF time or PoV space budget. Although that doesn't seem likely, I also feel that this simplifying assumption doesn't give us a lot - maybe it is not that hard to start with them?

Then, regarding downward messages. In the current version that we are discussing here, downward messages are implemented the same way you (@bkchr) introduced them: i.e. instead of a watermark we just used a simple `processed_downward_messages`. I liked this approach because, well, I didn't see a lot of reasons to use watermarks, but then today I finally broke down and switched to watermarks in https://github.com/paritytech/polkadot/pull/1629 (see the PR description to find out why).

But now I think maybe it would be a good idea to revert it back to `processed_downward_messages`. I'll think about that.

Reply:
Thank you for the explanation :) 

> Although that doesn't seem likely, I also feel that this simplifying assumption doesn't give us a lot - maybe it is not that hard to start with them?

I'm still not sure that this is true, IMHO we can reach this limit relative easy. In the end the parachain wants to process some of its own transactions as well. Just alone think about a parathread that collects messages over quite some time, the parachain will probably not be able to process them in one block.

Reply:
I think in a big part it depends on the parametrization and there are two parameters in play: weight and pov size. I don't think that it is productive to speak about the weight here, because it is highly dependent on the exact code of PVF at hand and more primarily it's PVF which will be designed to not violate any constraints put onto the it by polkadot (although we need to come up with a less costraining system overall, about this later). Therefore, let's assume for now that messages received will always fit into the weight constraints.

Regarding the PoV size, this could be enforced simply. 

Since: 

1. there can be only so many paras, `N`, executed at the same time, and
1. every parachain can target the same single parachain, and
1. every parachain can only send one XCMP/HRMP message per block

we arrive to the worst case of `N` messages can be received at the same time, then we can say that we want to devote a half of PoV at most for messages. Dividing the terms gives us the maximum size of the message which would make it impossible to overwhelm the parachain as long as it's PVF doesn't use more than half of its PVF (and once again, it should follow the constraints unless it accepts possibility of getting in the aforementioned unfortunate situation).


Now, regarding the parathread, I don't think they are too special here

Reply:
>     3\. every parachain can only send one XCMP/HRMP message per block

Okay, I never heard from this before. However, this clearly lowers the amount of messages :)

Reply:
I beg your pardon! It is actually

> every parachain can only send one XCMP/HRMP message per block **per recipient**